In [ ]:
Note: training and test data wrangled, can do first itteration of Modeling WITH test data.

    -after first pass modeling, CONVERT MFCC features to 40 from 13.
    -also, go back to classes labels and consider reducing number of classes.

# Wrangling: Extracting Spectrograms, MFCC Spectrograms, Mean-MFCC
## (modeling in next notebook)

In [ ]:
import pandas as pd


In [2]:
import pandas as pd
import numpy as np
import time
import librosa
import librosa.display
import IPython.display as ipd
import matplotlib.pyplot as plt
from PIL import Image
import skimage
import skimage.io
import os
import pickle
import keras
from keras.models import Sequential
from keras.layers import Dense, Dropout, Flatten, Conv2D, MaxPooling2D
from keras.utils import to_categorical
from keras import optimizers

%matplotlib inline

Using TensorFlow backend.
/Users/Mateo/opt/anaconda3/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:516: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
/Users/Mateo/opt/anaconda3/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:517: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
/Users/Mateo/opt/anaconda3/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:518: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
/Users/Mateo/opt/anaconda3/lib/python3.7/sit

In [3]:
dev_info = pd.read_json('data/labelled_dev_info.json')


In [6]:
eval_info = pd.read_json('data/labelled_eval_info.json')

## Converting Spectrograms and saving as Image Files

In [ ]:
# Not scaling for now...

# #function for scaling spectrogram images (however, the datagenerator can do that too, so maybe unnecessary here)

# def scale_minmax(X, min=0.0, max=1.0):
#     X_std = (X - X.min()) / (X.max() - X.min())
#     X_scaled = X_std * (max - min) + min
#     return X_scaled

In [14]:
base_dir = '/Users/Mateo/Springboard/FSD50k/'
audio_dir = '/Users/Mateo/Springboard/FSD50k/data/FSD50K.dev_audio/'
test_audio_dir = '/Users/Mateo/Springboard/FSD50k/data/FSD50K.eval_audio/'
list_of_wav_names = dev_info.wav_name.to_list()
test_wav_names_lst = eval_info.wav_name.to_list()

### Spectorgrams for Training Data

In [32]:
# # Spectorgrams for Training Data

# # creating directory
# os.chdir(base_dir)
# if not os.path.exists('data/spec_50k'):
#     os.makedirs('data/spec_50k')
    
# # changing into desired directory to save spectrogram images
# path = 'data/spec_50k'
# os.chdir(path)


# start = time.time()

# for i in range(len(list_of_file_names)):
#     file_path = audio_dir + list_of_file_names[i]
#     data, sr = librosa.load(file_path)  # consider using 'kaiser_fast' instead of 'kaiser_best' for faster load time
                                          # data, sr = librosa.load(file_path, res_type='kaiser_fast')
    
#     # trimming to 5 seconds
#     # padding with random offset for shorter tracks to 5 seconds

#     input_length = 5 * 22050 # 5 seconds * 22050 samples per second (the sample_rate)

#     if len(data) > input_length:
#         data = data[:input_length]

#     elif input_length > len(data):
#         max_offset = input_length - len(data)
#         offset = np.random.randint(max_offset)
#         data = np.pad(data, (offset, input_length - len(data) - offset), "constant")
    
#     # Fast Fourier Transform, a window for the results image
#     n_fft = 2048
#     # hop length slides the window (4:1 something er other?)
#     hop_length = 512
#     # converts audio spectrum into 128 evenly spaced groups based on human hearing
#     n_mels = 128

#     S = librosa.feature.melspectrogram(data, sr=sr, n_fft=n_fft, hop_length=hop_length, n_mels=n_mels)
#     S_DB = librosa.power_to_db(S, ref=np.max)
    
    
#     S_DB = S_DB.astype(np.uint8) #converting from float32 to uint8 (more efficient file format)
    
#     # NOT SCALING IN THIS ITERATION....
# #     # # min-max scale to fit inside 8-bit range

# #     img = scale_minmax(S_DB, 0, 255).astype(np.uint8)
# #     img = np.flip(img, axis=0) # put low frequencies at the bottom in image
# #     img = 255-img # invert. make black==more energy

#     skimage.io.imsave(list_of_file_names[i][:-4] + '.png', S_DB)  # removing .wav suffix with .png 
#     # to display impage: librosa.display.specshow(img, sr=sr, hop_length=hop_length)
    
# end = time.time()
# print(end - start)

/Users/Mateo/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:53: UserWarning: 55356.png is a low contrast image
/Users/Mateo/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:53: UserWarning: 95076.png is a low contrast image
/Users/Mateo/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:53: UserWarning: 125660.png is a low contrast image
/Users/Mateo/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:53: UserWarning: 164482.png is a low contrast image
/Users/Mateo/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:53: UserWarning: 166707.png is a low contrast image
/Users/Mateo/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:53: UserWarning: 166709.png is a low contrast image
/Users/Mateo/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:53: UserWarning: 176970.png is a low contrast image
/Users/Mateo/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:53: UserWarning: 33367

13205.982905864716


### Spectorgrams for Test Data


In [17]:
# Spectorgrams for Test Data

# creating directory
os.chdir(base_dir)
if not os.path.exists('data/test_spec_50k'):
    os.makedirs('data/test_spec_50k')
    
# changing into desired directory to save spectrogram images
path = 'data/test_spec_50k'
os.chdir(path)


start = time.time()

for i in range(len(test_wav_names_lst)):
    file_path = test_audio_dir + test_wav_names_lst[i]
    # consider using 'kaiser_fast' instead of 'kaiser_best' for faster load time
    data, sr = librosa.load(file_path, res_type='kaiser_fast')  
    
    # trimming to 5 seconds
    # padding with random offset for shorter tracks to 5 seconds

    input_length = 5 * 22050 # 5 seconds * 22050 samples per second (the sample_rate)

    if len(data) > input_length:
        data = data[:input_length]

    elif input_length > len(data):
        max_offset = input_length - len(data)
        offset = np.random.randint(max_offset)
        data = np.pad(data, (offset, input_length - len(data) - offset), "constant")
    
    # Fast Fourier Transform, a window for the results image
    n_fft = 2048
    # hop length slides the window (4:1 something er other?)
    hop_length = 512
    # converts audio spectrum into 128 evenly spaced groups based on human hearing
    n_mels = 128

    S = librosa.feature.melspectrogram(data, sr=sr, n_fft=n_fft, hop_length=hop_length, n_mels=n_mels)
    S_DB = librosa.power_to_db(S, ref=np.max)
    
    
    S_DB = S_DB.astype(np.uint8) #converting from float32 to uint8 (more efficient file format)
    skimage.io.imsave(test_wav_names_lst[i][:-4] + '.png', S_DB)  
    
end = time.time()
print(end - start)

/Users/Mateo/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:53: UserWarning: 387216.png is a low contrast image
/Users/Mateo/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:53: UserWarning: 38709.png is a low contrast image
/Users/Mateo/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:53: UserWarning: 366738.png is a low contrast image
/Users/Mateo/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:53: UserWarning: 7417.png is a low contrast image
/Users/Mateo/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:53: UserWarning: 110580.png is a low contrast image


950.5249390602112


Spectrogram: Total run time of cell above: 13205.982905864716 (3.5 hours!) (training)

Spectrogram: Test took 15 minutes, why so much faster (kaiser_fast?)

# Extracting MFCC Spectrogram ( Convert to 40 MFCC Features at some point)
## with this next iteration, extract Mean MFCC values before converting to np.uint8

In [38]:
# # saving MFCC's to disk in mfcc_50k directory

# # creating directory
# os.chdir(base_dir)
# if not os.path.exists('data/mfcc_50k'):
#     os.makedirs('data/mfcc_50k')
    
# # changing into desired directory to save mfcc images
# path = 'data/mfcc_50k'
# os.chdir(path)



# start = time.time()

# for i in range(len(list_of_file_names)):
#     file_path = audio_dir + list_of_file_names[i]
#     data, sr = librosa.load(file_path)
    
#     #padding/trimming to 5 seconds with random offset for shorter tracks

#     input_length = 5 * 22050

#     if len(data) > input_length:
#         data = data[:input_length]

#     elif input_length > len(data):
#         max_offset = input_length - len(data)
#         offset = np.random.randint(max_offset)
#         data = np.pad(data, (offset, input_length - len(data) - offset), "constant")
    
#     # extracting mfcc features
#     mfcc = librosa.feature.mfcc(y=data, sr=sr, n_mfcc=13)
    
#     # converting to uint8 num_type
#     mfcc = mfcc.astype(np.uint8)
#     skimage.io.imsave(list_of_file_names[i][:-4] + '.png', mfcc)  # removing .wav suffix with .png 

    
# end = time.time()
# print(end - start)

36347.03870201111


In [42]:
print('MFCC Spectrogram Time:', 36347, '10 hours to run!. I think things were being stored in memory while running...')

MFCC Spectrogram Time: 36347 10 hours to run!. I think things were being stored in memory while running...


### Extracting Test Data MFCC Spectrogram and Mean MFCC Values

In [20]:
# saving Test Data MFCC's to disk in mfcc_50k directory

# creating directory
os.chdir(base_dir)
if not os.path.exists('data/test_mfcc_50k'):
    os.makedirs('data/test_mfcc_50k')
    
# changing into desired directory to save mfcc images
path = 'data/test_mfcc_50k'
os.chdir(path)



start = time.time()



# intanitating numpy array for Mean MFCC values
num_files = len(test_wav_names_lst)
num_mfcc_features = 13
mfcc_mean_features = np.zeros(shape=(num_files, num_mfcc_features))



for i in range(len(test_wav_names_lst)):
    file_path = test_audio_dir + test_wav_names_lst[i]
    data, sr = librosa.load(file_path, res_type = 'kaiser_fast')
    
    #padding/trimming to 5 seconds with random offset for shorter tracks

    input_length = 5 * 22050

    if len(data) > input_length:
        data = data[:input_length]

    elif input_length > len(data):
        max_offset = input_length - len(data)
        offset = np.random.randint(max_offset)
        data = np.pad(data, (offset, input_length - len(data) - offset), "constant")
    
    # extracting mfcc features
    mfcc = librosa.feature.mfcc(y=data, sr=sr, n_mfcc=13)
    
    
    # extracting Mean for each MFCC band
    mfcc_processed = np.mean(mfcc, axis= 1)  # orginally: np.mean(foo.T, axis= 0), but same no?
    # appending mean values to features list
    mfcc_mean_features[i] = mfcc_processed
    
    # converting to uint8 num_type
    mfcc = mfcc.astype(np.uint8)
    skimage.io.imsave(test_wav_names_lst[i][:-4] + '.png', mfcc)  # removing .wav suffix with .png 
    

# saving numpy array to disk    
mean_features_filepath = base_dir + 'data/test_mfcc_mean_features.npz'
np.savez(mean_features_filepath, mfcc_mean_features)    

end = time.time()
print(end - start)

992.9505019187927


In [21]:
992/60


16.533333333333335

Test MFCC Spectrogram and Mean MFCC ran in 16 minutes (all cuz kaiser_fast, or maybe I closed some tabs....)

## Converting Training MFCC to Mean MFCC ( Convert to 40 MFCC Features at some point)

In [85]:
# # Creating Numpy Matrix of Mean MFCC Features for Training Data




# mfcc_dir = '/Users/Mateo/Springboard/FSD50k/data/mfcc_50k/'

# num_files = len(list_of_file_names)
# num_mfcc_features = 13

# mfcc_mean_features = np.zeros(shape=(num_files, num_mfcc_features))

# start = time.time()

# for i in range(len(list_of_file_names)):  
#     file_path = mfcc_dir + list_of_file_names[0][:-4] + '.png'
    
#     # loading processed MFCC Spectrogram Images
#     mfcc = skimage.io.imread(file_path)

#     # extract mfcc mean for each MFCC band
#     mfcc_processed = np.mean(mfcc, axis= 1)  # orginally: np.mean(foo.T, axis= 0), but same no?

#     # appending mean values to features list
#     mfcc_mean_features[i] = mfcc_processed
    
# end = time.time()
# print(end - start)

18.719588041305542


In [94]:
mean_features_filepath = base_dir + 'data/mfcc_mean_features.npz'
np.savez(mean_features_filepath, mfcc_mean_features)

In [103]:
# loading npz for sanity check   # works!

loaded_mfcc_mean_features = np.load(mean_features_filepath)['arr_0']

In [107]:
mfcc_mean_features.shape

(40966, 13)

In [106]:
loaded_mfcc_mean_features.shape

(40966, 13)

# NOTES BELOW